sdmetrics/reports/single_table/_properties/column_pair_trends.py

```diff
@@ -0,0 +1,441 @@
+            error = self._columns_datetime_conversion_failed[column_name_2]
+
+        elif self._sdtype_to_shape[sdtype_col_1] != self._sdtype_to_shape[sdtype_col_2]:
+            if column_name_1 in self._columns_discretization_failed.keys():
+                error = self._columns_discretization_failed[column_name_1]
+            elif column_name_2 in self._columns_discretization_failed.keys():
+                error = self._columns_discretization_failed[column_name_2]
+
+        return error
+
+    def _generate_details(self, real_data, synthetic_data, metadata, progress_bar):
+        """Generate the _details dataframe for the column pair trends property.
+
+        Args:
+            real_data (pandas.DataFrame):
+                The real data
+            synthetic_data (pandas.DataFrame):
+                The synthetic data
+            metadata (dict):
+                The metadata of the table
+            progress_bar:
+                The progress bar to use. Defaults to tqdm.
+        """
+        processed_real_data = self._get_processed_data(real_data, metadata)
+        processed_synthetic_data = self._get_processed_data(synthetic_data, metadata)
+
+        column_names_1 = []
+        column_names_2 = []
+        metric_names = []
+        scores = []
+        real_correlations = []
+        synthetic_correlations = []
+        error_messages = []
+
+        list_dtypes = self._sdtype_to_shape.keys()
+        for column_names in itertools.combinations(list(metadata['columns']), r=2):
+            column_name_1 = column_names[0]
+            column_name_2 = column_names[1]
+
+            sdtype_col_1 = metadata['columns'][column_name_1]['sdtype']
+            sdtype_col_2 = metadata['columns'][column_name_2]['sdtype']
+
+            error = None
+            valid_sdtypes = sdtype_col_1 in list_dtypes and sdtype_col_2 in list_dtypes
+            if not valid_sdtypes:
+                if progress_bar:
+                    progress_bar.update()
+
+                continue
+
+            metric = self._get_metric(sdtype_col_1, sdtype_col_2)
+            try:
+                if self._required_preprocessing(sdtype_col_1, sdtype_col_2):
+                    error = self._preprocessing_failed(
+                        column_name_1, column_name_2, sdtype_col_1, sdtype_col_2
+                    )
+                    if error:
+                        raise Exception('Preprocessing failed')
+
+                columns_real = self._get_columns_data(
+                    column_name_1, column_name_2, processed_real_data, metadata
+                )
+                columns_synthetic = self._get_columns_data(
+                    column_name_1, column_name_2, processed_synthetic_data, metadata
+                )
+
+                score_breakdown = metric.compute_breakdown(
+                    real_data=columns_real, synthetic_data=columns_synthetic
+                )
```

Comment:
I think the logic in these three methods can be done in one method that just takes in the data, discretized data, column names and metadata. Then it can just check which metric to run and pick the columns from either the original or discretized data

Reply:
Yes I agree, done in [ebe9e54](https://github.com/sdv-dev/SDMetrics/pull/368/commits/ebe9e54c38ed06f78a6c4ca1a0bf16eb524501e7)